### PR TITLE
Gather all logged runs in a standard date format

### LINKED ISSUE #7: Add tests for functions that involve runner statistics
Due to some flaky behavior involving times with javascript and the GitHub Action that runs tests, tests for the following functions were removed in [#eb2ba2b](https://github.com/slack-samples/deno-virtual-running-buddies/pull/5/commits/eb2ba2b04d3ac542d251188098136e79b0ad23a6):

* `collect_runner_stats`
* `collect_team_stats`
* `format_leaderboard`

Here's an example of a [failing run](https://github.com/slack-samples/deno-virtual-running-buddies/actions/runs/4491541554) and [passing run](https://github.com/slack-samples/deno-virtual-running-buddies/actions/runs/4534385340) on the same commit... 😪 

These tests should ideally be returned, but in a more stable state! Possibly by using defined dates and times in the test? My current hypothesis is that these tests were passing or failing based on what time the action began, but I am somewhat unsure. Something to soon explore!

### LINKED ISSUE #13: Leaderboard workflow reports zero miles ran after logging a run
1. Create the Log a run trigger
2. Run the Log a run trigger
3. Create the Display Leaderboard trigger
4. Run the Display leaderboard trigger
5. Expect the leaderboard to correctly reflect the number of miles you ran

Actual result: 0 miles ran

cc @e-zim

### LINKED ISSUE #22: Leaderboard workflow has errors in date math and query assembly
When I ran the leaderboard workflow today, I see this:

```
2023-09-21 12:57:40 [info] [Wf056QRA24UW] (Trace=Tr05TADFAYDT) Shortcut trigger successfully started execution of function 'Display the leaderboard'
2023-09-21 12:57:40 [info] [Fn056UFF79GD] (Trace=Tr05TADFAYDT) Function 'Display the leaderboard' (workflow function) started
2023-09-21 12:57:41 [info] [Wf056QRA24UW] (Trace=Tr05TADFAYDT) Workflow 'Display the leaderboard' started
2023-09-21 12:57:41 [info] [Wf056QRA24UW] (Trace=Tr05TADFAYDT) Workflow step 1 of 4 started
2023-09-21 12:57:41 [info] [Fn056ZU6R6DA] (Trace=Tr05TADFAYDT) Function 'Collect team stats' (app function) started
2023-09-21 12:57:42 [info] [Datastore:running_datastore] (Trace=Tr05TADFAYDT) Datastore query succeeded with '{"expression":"#date BETWEEN :start_date AND :end_date","expression_attributes":{"#date":"rundate"},"expression_values":{":start_date":"9\/15\/2023",":end_date":"9\/21\/2023"},"limit":100,"cursor":""}'
2023-09-21 12:57:42 [error] [Datastore:running_datastore] (Trace=Tr05TADFAYDT) Datastore query failed with '{"expression":"#date BETWEEN :start_date AND :end_date","expression_attributes":{"#date":"rundate"},"expression_values":{":start_date":"9\/8\/2023",":end_date":"9\/14\/2023"},"limit":100,"cursor":""}'
	ValidationException: Invalid FilterExpression: The BETWEEN operator requires upper bound to be greater than or equal to lower bound; lower bound operand: AttributeValue: {S:9/8/2023}, upper bound operand: AttributeValue: {S:9/14/2023}
2023-09-21 12:57:42 [error] [Fn056ZU6R6DA] (Trace=Tr05TADFAYDT) Function 'Collect team stats' (%!!(MISSING)s(<nil>) function) failed
	Failed to retrieve past runs: datastore_error
2023-09-21 12:57:43 [error] [Wf056QRA24UW] (Trace=Tr05TADFAYDT) Workflow step 'Collect team stats' failed
2023-09-21 12:57:43 [error] [Wf056QRA24UW] (Trace=Tr05TADFAYDT) Workflow 'Display the leaderboard' failed
	Failed to retrieve past runs: datastore_error
2023-09-21 12:57:43 [error] [Fn056UFF79GD] (Trace=Tr05TADFAYDT) Function 'Display the leaderboard' (%!!(MISSING)s(<nil>) function) failed
	Failed to retrieve past runs: datastore_error
```

The error message is correct: the following are not ordered correctly:

```
ValidationException: Invalid FilterExpression: The BETWEEN operator requires upper bound to be greater than or equal to lower bound; lower bound operand: AttributeValue: {S:9/8/2023}, upper bound operand: AttributeValue: {S:9/14/2023}
```

DynamoDB is reporting that the values being compared, start date and end date, are not ordered. Since these are stored as strings (that's what the `{S:}` in the error stands for), Dynamo will want these to be sorted lexicographically. This is a problem for our sample implementation because we are using American dates, which are not sorted lexicographically 😬.

So we should fix this. [Dynamo recommends to use ISO-8601 date strings to store dates and times](https://aws.amazon.com/blogs/database/working-with-date-and-timestamp-data-types-in-amazon-dynamodb/). We could use a number, too, but the nice thing about ISO strings is that they are human readable.

### PR DESCRIPTION
**Pull request by zimeg**
### Type of change

- [ ] New sample
- [x] New feature
- [x] Bug fix
- [ ] Documentation

### Summary

This PR paginates over the `running_datastore` datastore to collect all logged runs and stores runs with with the ISO-8601 timestamp. Tests are also returned with a stable date to verify proper month/year wrapping when gathering past weeks.

Closes #7, fixes #22, and resolved #13.

### Requirements

- [x] I’ve checked my submission against the Samples Checklist to ensure it complies with all standards
- [x] I have ensured the changes I am contributing align with existing patterns and have tested and linted my code
- [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct)

### CODE REVIEW

**Review by misscoded — APPROVED** [review]

**Comment by zimeg** [comment]
Thanks for the review @misscoded! 🚀

File: datastores/run_data.ts [hunk 1]
```diff
@@ -21,4 +24,44 @@ const RunningDatastore = DefineDatastore({
...
+        ...expressions,
+      });
+
+    if (!runs.ok) {
+      return { ok: false, items, error: runs.error };
+    }
+
+    cursor = runs.response_metadata?.next_cursor;
+    items.push(...runs.items);
+  } while (cursor);
```
Comment (filmaj) [inline datastores/run_data.ts:62]: `do`/`while` w/ cursors in Dynamo is exactly how I structure my DDB pagination calls 👍

File: functions/collect_runner_stats.ts [hunk 2]
```diff
@@ -24,28 +25,27 @@ export const CollectRunnerStatsFunction = DefineFunction({
...
+  if (!runs.ok) {
+    return { error: `Failed to retrieve past runs: ${runs.error}` };
+  }
 
   // Add run statistics to the associated runner
   runs.items.forEach((run) => {
-    const isRecentRun = run.rundate >=
-      startOfLastWeek.toLocaleDateString("en-CA", { timeZone: "UTC" });
+    const isRecentRun =
+      run.rundate >= startOfLastWeek.toISOString().substring(0, 10);
```
Comment (filmaj) [inline functions/collect_runner_stats.ts:48]: 👨‍🍳 💋

**Review by filmaj — COMMENTED** [review]
Nicely done!

File: functions/collect_runner_stats.ts (hunk 2 above)
Comment (zimeg) [reply functions/collect_runner_stats.ts:48]: ISO-8601 just feels so much better 🙌 🙌

Comment (filmaj) [reply functions/collect_runner_stats.ts:48]: ISO8601 is the best date standard IMO

**Pull request merged**